### ISSUE TITLE
Is reading / loading large TIFFs and PSDs inherently slow?

### ISSUE THREAD

**Issue by octavian-nita** [haraldk/TwelveMonkeys#97]
Hi guys,

I am using TwelveMonkeys to put together a small proof-of-concept image viewer. As one can see in the status bar info from the first screenshot [here](https://github.com/octavian-nita/thepholio-desktop/releases), it takes a while (on average, around a second) to load a medium-sized (I would say) TIFF file (the computers I use are "quite fast", including a MacBook Pro so I will skip the configuration details for the moment; I am fairly sure it happens the same on many computers).

The displayed time is only measuring the loading of a file to a BufferedImage and not its rendering on the screen, something along the lines:

```
stopwatch.reset().start();
BufferedImage image = Images.load(input);
long millisLoaded = stopwatch.stop().elapsed();
```

where [`Images.load(input)`](https://github.com/octavian-nita/thepholio-desktop/blob/master/src/main/java/org/thepholio/image/Images.java) is mostly implemented following the TwelveMonkeys documentation and trying to minimize memory consumption:

```
    public static BufferedImage load(Object imageInput) throws FormatNotSupported, IOException {
        try (ImageInputStream input = ImageIO.createImageInputStream(imageInput)) {

            Iterator<ImageReader> readers = ImageIO.getImageReaders(input);
            if (!readers.hasNext()) {
                throw imageInput instanceof File ? new FormatNotSupported((File) imageInput)
                                                 : new FormatNotSupported("Input image format not supported");
            }

            ImageReader reader = readers.next();
            try {
                reader.setInput(input);

                BufferedImage image =
                    GRAPHICS_CONFIGURATION.createCompatibleImage(reader.getWidth(0), reader.getHeight(0));
                ImageTypeSpecifier imageSpecifier = ImageTypeSpecifier.createFromBufferedImageType(image.getType());

                boolean canReadDirectly = false;
                for (Iterator<ImageTypeSpecifier> readerSpecifiers = reader.getImageTypes(0);
                     readerSpecifiers.hasNext(); ) {
                    if (imageSpecifier.equals(readerSpecifiers.next())) {
                        canReadDirectly = true;
                        break;
                    }
                }

                if (canReadDirectly) {
                    ImageReadParam param = reader.getDefaultReadParam();
                    param.setDestination(image);
                    reader.read(0, param);
                } else {
                    Graphics graphics = image.getGraphics();
                    graphics.drawImage(reader.read(0), 0, 0, null);
                    graphics.dispose();
                }

                return image;
            } finally {
                reader.dispose();
            }
        }
    }
```

Similar issues I get when loading PSD files (although it seems to be a bit faster).

## TL;DR

Now, while I understand there's a lot to read and transform / interpret from such a file and that our technology of choice (i.e. Java), despite being quite good for most things is still not amazing when it comes to image manipulation, I was wondering if there's anything I could do (either in how I use TwelveMonkeys or in my algorithm) to increase the loading speed of such large images (especially TIFFs and PSDs).

Thanks and regards,
Octavian

**Comment by haraldk** [comment]
Hi Tavi (can I call you that? :-),

I'm sure there are things that can be done to the `ImageReader` plugins themselves, to make them faster. At the moment though, my priority is format support over performance. Not that I'm not trying to implement things as fast as possible, just that I haven't spent quality time with the profiler to see where the big gains can be made.

If you have the time to help here, I think many people would be happy! :-)

Some of the factors that matters here:
 - Raw I/O speed. In this regard Java should be quite on par with the best. 
 - Decompression algorithms. I've implemented PackBits, Huffman, LZW etc. from scratch, they are likely not the fastest implementations out there. 
 - Java array handling and integer arithmetic. There's loads of array copying going on, re-shuffling and re-arranging to get the bits in the order standard `BufferedImage`s like.

Some things I've discovered during many years of ImageIO development:
- `ImageIO.setUseCache(false);` will turn off disk caching (which is enabled by default), always makes things faster. 
- What kind of input you are using matters. From my testing, using `File` (resulting in a `RAFImageInputStream` is usually the fastest). I've also experimented with using nio stuff for reading, but without the gains I was hoping for. 

PS: Your timing includes drawing/conversion time, it would be interesting to split reading and conversion time. 

-- 
Harald K

**Comment by haraldk** [comment]
Hi again,

Did some more testing, the main reason reading the TIFFs (body-01.tif and body-02.tif) are slow, is due to conversion, as the they are RGB 16 bit/sample, and has to be converted to normal RGB 8 bit/sample to be compatible. The read time only on my computer is < 200ms (after initial load). It could be possible to read these images in 8 bit only (discard the lower 8 bits) for better performance, if only 8 bits/sample was required. It could potentially save 40-60% of the time, but I haven't looked into that yet. 

For the large PSD (home-away-from-home.psd) read time is < 300ms (including the Adobe to sRGB conversion, which alone takes about 140ms). Could be made faster without the conversion, but then the colors would be dull (again)... 

The images are all uncompressed, so decompression speed is not a factor here.

-- 
Harald K

**Comment by octavian-nita** [comment]
Hi Harald,

>> Hi Tavi (can I call you that? :-),
Please do! :)

Thanks for shedding a bit more light on this. Actually I had thought about helping out but I didn't exactly know where to start. I guess a bit of debugging and profiling would be appropriate.

I will try to follow (i.e. debug) what you said about reading tiffs "for compatible devices".

Btw, I don't have much experience with digital image representation and I imagine there are many resources out there but... would you recommend any to start with?

Thanks, Tavi

**Comment by haraldk** [comment]
Hmmm..

That depends on what you want to learn I guess. 

I had a course in university (almost 20 years ago), and the one book I remember as useful for learning the basics is [Computer Graphics: Principles and Practice](http://en.wikipedia.org/wiki/Computer_Graphics:_Principles_and_Practice), but I'm not sure how up to date it is now...

Reading the specs for the [TIFF](https://partners.adobe.com/public/developer/en/tiff/TIFF6.pdf) and [PSD](http://www.adobe.com/devnet-apps/photoshop/fileformatashtml/) formats, is of course also vital to understand those formats.

And finally, there's a lot to learn from reading the source code of others. I've spent a lot of time reading the source code from the JDK, to better understand how to use certain classes, or how certain interfaces/methods are best implemented.


Best regards,

-- 
Harald K

**Comment by haraldk** [comment]
Closing this for now. 

Feel free to reopen if you have further questions.

-- 
Harald K
